course05/homework02/freekpaans/linked_list.h

```diff
@@ -0,0 +1,166 @@
+#include <cstddef>
+#include <memory>
+#include <iterator>
+#include <numeric>
+
+template <typename T>
+class LinkedList {
+public:
+	using size_type = std::size_t;
+	using value_type = T;
+	using reference = T&;
+	using const_reference = const T&;
+
+	struct Link
+	{
+		T mValue;
+		std::unique_ptr<Link> mNext;
+	};
+
+	using LinkPtr = std::unique_ptr<Link>;
+
+	LinkedList() =default;
+	LinkedList(const LinkedList& from);
+
+	LinkedList& operator=(const LinkedList& from);
+
+	bool empty() const
+	{
+		return !mFront;
+	}
+
+	size_type size() const;
+
+	void push_front(const_reference str);
+	void pop_front();
+	void insert_at(size_type index, const_reference str);
+
+	const_reference front() const;
+	const_reference at(size_type index) const;
+
+	struct iterator : public std::iterator<std::input_iterator_tag, T, size_type>
+	{
+		const Link* mLink;
+
+		explicit iterator(const Link* link) : mLink{link}
+		{}
+
+		bool operator==(const iterator& other) const
+		{
+			return mLink == other.mLink;
+		}
+
+		bool operator!=(const iterator& other) const
+		{
+			return !(*this == other);
+		}
+
+		iterator& operator++()
+		{
+			mLink = mLink->mNext.get();
+			return *this;
+		}
+
+		const_reference operator*() const
+		{
+			return mLink->mValue;
+		}
+	};
+
+	iterator begin() const
+	{
+		return iterator{mFront.get()};
+	}
+
+	iterator end() const
+	{
+		return iterator{nullptr};
+	}
+
+private:
+	LinkPtr mFront{};
+};
+
+template <typename T>
+static void push_at(
+	typename LinkedList<T>::LinkPtr& link,
```

Comment:
This is one instance for the verbosity of `typename` where it concerns a standalone function. If I make it a static member of `LinkedList` this would go away I think. Or are there any other solutions?